File: src/features.cpp
#include "features.hpp"

#include <algorithm>
#include <cmath>
#include <map>
#include <numeric>
#include <random>
#include <stdexcept>

#include "config.hpp"

namespace era {
namespace {

// pandas skips NaN in .mean(); an all-NaN group yields NaN, which is what lets
// the caller fall through to the next .fillna() in the chain.
double nanmean(const std::vector<double>& values) {
    double total = 0.0;
    std::size_t count = 0;
    for (const double v : values) {
        if (!std::isnan(v)) {
            total += v;
            ++count;
        }
    }
    return count ? total / static_cast<double>(count) : std::nan("");
}

// Group mean broadcast back over rows, i.e. df.groupby(key)[col].transform("mean").
std::vector<double> group_mean_transform(const std::vector<std::string>& keys,
                                         const std::vector<double>& values) {
    std::map<std::string, std::pair<double, std::size_t>> sums;
    for (std::size_t i = 0; i < values.size(); ++i) {
        if (!std::isnan(values[i])) {
            auto& entry = sums[keys[i]];
            entry.first += values[i];
            entry.second += 1;
        }
    }
    std::vector<double> out(values.size());
    for (std::size_t i = 0; i < values.size(); ++i) {
        auto it = sums.find(keys[i]);
        out[i] = (it != sums.end() && it->second.second)
                     ? it->second.first / static_cast<double>(it->second.second)
                     : std::nan("");
    }
    return out;
}

const std::vector<double>& column(const Frame& df, const std::string& name) {
    auto it = df.num.find(name);
    if (it == df.num.end()) {
        throw std::runtime_error("missing column: " + name);
    }
    return it->second;
}

}  // namespace

std::string decade_of(const std::string& season_year) {
    const int start = std::stoi(season_year.substr(0, 4));
    return std::to_string(start / 10 * 10) + "s";
}

std::vector<double> compute_age(const Frame& df) {
    const std::size_t n = df.size();
    std::vector<double> age(n, std::nan(""));
    std::vector<std::string> decade(n);
    for (std::size_t i = 0; i < n; ++i) {
        decade[i] = decade_of(df.season_year[i]);
        const std::string& birth = df.birth_date[i];
        if (birth.size() >= 4 && std::all_of(birth.begin(), birth.begin() + 4,
                                             [](unsigned char c) { return std::isdigit(c); })) {
            age[i] = std::stoi(df.season_year[i].substr(0, 4)) - std::stoi(birth.substr(0, 4));
        }
    }

    const std::vector<double> dec_mean = group_mean_transform(decade, age);
    const double overall_mean = nanmean(age);
    for (std::size_t i = 0; i < n; ++i) {
        if (std::isnan(age[i])) {
            age[i] = dec_mean[i];
        }
        if (std::isnan(age[i])) {
            age[i] = overall_mean;
        }
        if (std::isnan(age[i])) {
            age[i] = 0.0;
        }
    }
    return age;
}

std::vector<double> build_value_target(const Frame& df) {
    const std::size_t n = df.size();
    std::vector<double> raw(n, 0.0);
    for (const std::string& col : VALUE_ALWAYS) {
        const std::vector<double>& values = column(df, col);
        for (std::size_t i = 0; i < n; ++i) {
            raw[i] += std::isnan(values[i]) ? 0.0 : values[i];
        }
    }
    for (const auto& [col, weight] : VALUE_CONDITIONAL) {
        const std::vector<double>& values = column(df, col);
        for (std::size_t i = 0; i < n; ++i) {
            raw[i] += (std::isnan(values[i]) ? 0.0 : values[i]) * weight;
        }
    }

    // Per-season z-score with ddof=0; a zero-variance season maps to all zeros.
    std::map<std::string, std::vector<std::size_t>> by_season;
    for (std::size_t i = 0; i < n; ++i) {
        by_season[df.season_year[i]].push_back(i);
    }
    std::vector<double> out(n, 0.0);
    for (const auto& [season, rows] : by_season) {
        double mean = 0.0;
        for (const std::size_t i : rows) {
            mean += raw[i];
        }
        mean /= static_cast<double>(rows.size());
        double variance = 0.0;
        for (const std::size_t i : rows) {
            variance += (raw[i] - mean) * (raw[i] - mean);
        }
        const double std_dev = std::sqrt(variance / static_cast<double>(rows.size()));
        for (const std::size_t i : rows) {
            out[i] = std_dev > 0.0 ? (raw[i] - mean) / std_dev : 0.0;
        }
    }
    return out;
}

std::map<std::string, std::vector<double>> build_position_oh(const Frame& df) {
    std::map<std::string, std::vector<double>> out;
    for (const std::string& position : POSITIONS) {
        out["pos_" + position] = std::vector<double>(df.size(), 0.0);
    }
    for (std::size_t i = 0; i < df.size(); ++i) {
        const std::string& raw = df.position[i];
        const bool known = std::find(POSITIONS.begin(), POSITIONS.end(), raw) != POSITIONS.end();
        out["pos_" + (known && !raw.empty() ? raw : std::string("Unknown"))][i] = 1.0;
    }
    return out;
}

Frame engineer_features(const Frame& df) {
    Frame out = df;
    out.era.resize(out.size());
    for (std::size_t i = 0; i < out.size(); ++i) {
        out.era[i] = decade_of(out.season_year[i]);
    }
    out.num["age"] = compute_age(out);

    // Impute numeric features with decade mean (historical reality: stl/blk/3pt
    // not recorded early; advanced metrics absent pre-1970s). If an entire era
    // lacks a stat (fg3_pct pre-1979, stl/blk pre-1973, tov/net_rating/usg pre-1977),
    // the decade mean is NaN -> fall back to the global column mean.
    for (const std::string& col : NUMERIC_FEATURES) {
        std::vector<double>& values = out.num.at(col);
        const std::vector<double> dec_mean = group_mean_transform(out.era, values);
        const double global_mean = nanmean(values);
        for (std::size_t i = 0; i < values.size(); ++i) {
            if (std::isnan(values[i])) {
                values[i] = dec_mean[i];
            }
            if (std::isnan(values[i])) {
                values[i] = global_mean;
            }
        }
    }

    for (auto& [name, values] : build_position_oh(out)) {
        out.num[name] = std::move(values);
    }
    out.num["value_target"] = build_value_target(out);
    return out;
}

std::pair<std::vector<int64_t>, std::vector<int64_t>> stratified_split(
    const std::vector<int64_t>& indices, const std::vector<std::string>& strata, double test_frac,
    std::mt19937_64& rng) {
    std::map<std::string, std::vector<int64_t>> by_class;
    for (std::size_t i = 0; i < indices.size(); ++i) {
        by_class[strata[i]].push_back(indices[i]);
    }
    std::vector<int64_t> keep;
    std::vector<int64_t> held_out;
    for (auto& [label, members] : by_class) {
        std::shuffle(members.begin(), members.end(), rng);
        auto n_out = static_cast<std::size_t>(
            std::llround(static_cast<double>(members.size()) * test_frac));
        if (members.size() >= 2) {
            n_out = std::clamp<std::size_t>(n_out, 1, members.size() - 1);
        }
        held_out.insert(held_out.end(), members.begin(), members.begin() + n_out);
        keep.insert(keep.end(), members.begin() + n_out, members.end());
    }
    std::sort(keep.begin(), keep.end());
    std::sort(held_out.begin(), held_out.end());
    return {keep, held_out};
}

Splits make_splits(const Frame& df) {
    std::mt19937_64 rng(SEED);
    std::vector<int64_t> all(df.era.size());
    std::iota(all.begin(), all.end(), 0);

    auto [train, test] = stratified_split(all, df.era, TEST_FRAC, rng);

    std::vector<std::string> train_era(train.size());
    for (std::size_t i = 0; i < train.size(); ++i) {
        train_era[i] = df.era[train[i]];
    }
    auto [final_train, val] = stratified_split(train, train_era, VAL_FRAC, rng);

    return Splits{final_train, val, test};
}

std::vector<std::string> feature_columns(const Frame& df) {
    std::vector<std::string> cols = NUMERIC_FEATURES;
    cols.push_back("age");
    for (const std::string& position : POSITIONS) {
        cols.push_back("pos_" + position);
    }
    std::vector<std::string> missing;
    for (const std::string& col : cols) {
        if (!df.num.count(col)) {
            missing.push_back(col);
        }
    }
    if (!missing.empty()) {
        std::string message = "Missing feature columns:";
        for (const std::string& col : missing) {
            message += " " + col;
        }
        throw std::runtime_error(message);
    }
    return cols;
}

Scaler fit_scaler(const Frame& df, const std::vector<int64_t>& idx_train,
                  const std::vector<std::string>& cols) {
    Scaler scaler;
    for (const std::string& col : cols) {
        const std::vector<double>& values = column(df, col);
        double mean = 0.0;
        for (const int64_t i : idx_train) {
            mean += values[i];
        }
        mean /= static_cast<double>(idx_train.size());
        double variance = 0.0;
        for (const int64_t i : idx_train) {
            variance += (values[i] - mean) * (values[i] - mean);
        }
        const double std_dev = std::sqrt(variance / static_cast<double>(idx_train.size()));
        scaler.mean.push_back(mean);
        // StandardScaler leaves constant columns untouched rather than dividing by zero.
        scaler.scale.push_back(std_dev > 0.0 ? std_dev : 1.0);
    }
    return scaler;
}

Matrix transform_features(const Frame& df, const Scaler& scaler,
                          const std::vector<std::string>& cols) {
    Matrix out;
    out.rows = static_cast<int64_t>(df.size());
    out.cols = static_cast<int64_t>(cols.size());
    out.data.resize(static_cast<std::size_t>(out.rows) * out.cols);
    for (int64_t c = 0; c < out.cols; ++c) {
        const std::vector<double>& values = column(df, cols[c]);
        for (int64_t r = 0; r < out.rows; ++r) {
            out.data[r * out.cols + c] =
                static_cast<float>((values[r] - scaler.mean[c]) / scaler.scale[c]);
        }
    }
    return out;
}

std::vector<int64_t> era_codes(const Frame& df) {
    std::vector<int64_t> codes(df.size());
    for (std::size_t i = 0; i < df.size(); ++i) {
        const auto it = std::find(ERA_ORDER.begin(), ERA_ORDER.end(), df.era[i]);
        if (it == ERA_ORDER.end()) {
            throw std::runtime_error("unknown era: " + df.era[i]);
        }
        codes[i] = std::distance(ERA_ORDER.begin(), it);
    }
    return codes;
}

}  // namespace era

File: era_translator/features.py
"""Feature engineering: era labels, age, imputation, scaling, target, splits."""

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

from .config import (
    NUMERIC_FEATURES,
    SEED,
    TEST_FRAC,
    VAL_FRAC,
    VALUE_ALWAYS,
    VALUE_CONDITIONAL,
)

POSITIONS = ["G", "F", "C", "Unknown"]


def decade_of(season_year: str) -> str:
    """'2002-03' -> '2000s'."""
    start = int(str(season_year)[:4])
    return f"{start // 10 * 10}s"


def compute_age(df: pd.DataFrame) -> pd.Series:
    """Age at season start = season_end_year - birth_year; decade-mean fallback."""
    season_end = df["season_year"].str[:4].astype(int)
    birth = pd.to_datetime(df["birth_date"], errors="coerce").dt.year
    age = season_end - birth
    df_tmp = pd.DataFrame({"age": age, "decade": df["season_year"].map(decade_of)})
    dec_mean = df_tmp.groupby("decade")["age"].transform("mean")
    overall_mean = age.mean(skipna=True)
    return age.fillna(dec_mean).fillna(overall_mean).fillna(0.0)


def build_value_target(df: pd.DataFrame) -> pd.Series:
    """Per-season z-score of a box-score composite.

    raw = pts + reb + ast + (stl + blk - tov if recorded else 0)
    Then z-scored within each season_year so the target is era-normalized.
    """
    raw = pd.Series(0.0, index=df.index, dtype=np.float64)
    for col in VALUE_ALWAYS:
        raw = raw + df[col].fillna(0.0).astype(np.float64)
    for col, w in VALUE_CONDITIONAL.items():
        raw = raw + df[col].fillna(0.0).astype(np.float64) * w

    def zscore(s):
        std = s.std(ddof=0)
        return (s - s.mean()) / std if std > 0 else s * 0.0

    return raw.groupby(df["season_year"]).transform(zscore)


def build_position_oh(df: pd.DataFrame) -> pd.DataFrame:
    pos = df["position"].fillna("Unknown").where(
        df["position"].fillna("Unknown").isin(POSITIONS), "Unknown"
    )
    return pd.get_dummies(pos, prefix="pos").reindex(
        columns=[f"pos_{p}" for p in POSITIONS], fill_value=0
    ).astype(np.float32)


def engineer_features(df: pd.DataFrame) -> pd.DataFrame:
    """Adds era, age, position one-hot; returns a copy with FEATURES + TARGET + ERA columns."""
    out = df.copy()
    out["era"] = out["season_year"].map(decade_of)
    out["age"] = compute_age(out)

    # Impute numeric features with decade mean (historical reality: stl/blk/3pt
    # not recorded early; advanced metrics absent pre-1970s). If an entire era
    # lacks a stat (fg3_pct pre-1979, stl/blk pre-1973, tov/net_rating/usg pre-1977),
    # the decade mean is NaN -> fall back to the global column mean.
    for col in NUMERIC_FEATURES:
        dec_mean = out.groupby("era")[col].transform("mean")
        out[col] = out[col].fillna(dec_mean).fillna(out[col].mean())

    pos_oh = build_position_oh(out)
    out = pd.concat([out, pos_oh], axis=1)

    out["value_target"] = build_value_target(out)
    return out


def make_splits(df: pd.DataFrame):
    """Stratified train/val/test split by era. Returns (train, val, test) indices."""
    era_series = df["era"].astype("category").cat.codes
    idx = np.arange(len(df))
    idx_train, idx_test = train_test_split(
        idx, test_size=TEST_FRAC, stratify=era_series, random_state=SEED
    )
    train_era = era_series.iloc[idx_train]
    idx_train, idx_val = train_test_split(
        idx_train, test_size=VAL_FRAC, stratify=train_era, random_state=SEED
    )
    return idx_train, idx_val, idx_test


def feature_columns(df: pd.DataFrame) -> list:
    cols = list(NUMERIC_FEATURES) + ["age"] + [f"pos_{p}" for p in POSITIONS]
    missing = [c for c in cols if c not in df.columns]
    if missing:
        raise KeyError(f"Missing feature columns: {missing}")
    return cols


def fit_scaler(df: pd.DataFrame, idx_train, cols):
    scaler = StandardScaler()
    scaler.fit(df.loc[idx_train, cols])
    return scaler


def transform_features(df: pd.DataFrame, scaler, cols) -> np.ndarray:
    return scaler.transform(df[cols]).astype(np.float32)
